### PR TITLE
core: dispatch `ContractNegotiation` events

### LINKED ISSUE #1434: Event Framework: dispatch ContractNegotiation events
# Feature Request

The EDC should dispatch events regarding `ContractNegotiation` lifecycle as:
- `initialized`
- `requested`
- `offered`
- `approved`
- `declined`
- `confirmed`
- `failed`

The feature must follow the guidelines described in the [event-framework decision record](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/6678501fe0dca78d5727adfbe0ac1cad9310c9a1/docs/developer/decision-records/2022-06-03-event-framework)

## Which Areas Would Be Affected?
core

## Why Is the Feature Desired?
to permit asynchronous notifications

## Solution Proposal
-

## Type of Issue
new feature

## Checklist

- [x] assigned appropriate label?
- [x] **Do NOT select a milestone or an assignee!**

### PR DESCRIPTION
**Pull request by ndr-brt**
## What this PR changes/adds

Dispatch `ContractNegotiation` events

## Why it does that

To permit asynchronous notifications

## Further notes
- rename the event listeners in `<entity name>EventListener`
- Add a `getFailureDetail()` method on the `AbstractResult` class, that provides a join of all the failure messages without having to do it every time
- Bump awaitility to 4.2.0

## Linked Issue(s)

Closes #1434

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/listener/ContractNegotiationEventListener.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    private final EventRouter eventRouter;
+    private final Clock clock;
+
+    public ContractNegotiationEventListener(EventRouter eventRouter, Clock clock) {
+        this.eventRouter = eventRouter;
+        this.clock = clock;
+    }
+
+    @Override
+    public void initialized(ContractNegotiation negotiation) {
```
Comment (paullatzelsperger) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/listener/ContractNegotiationEventListener.java:40]: nit: shouldn't this be called `initiated`? At least that's the term we use in most other places.

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java [hunk 2]
```diff
@@ -162,6 +167,8 @@ private void registerServices(ServiceExtensionContext context) {
         CommandRunner<ContractNegotiationCommand> commandRunner = new CommandRunner<>(commandHandlerRegistry, monitor);
 
         var observable = new ContractNegotiationObservableImpl();
+        observable.registerListener(new ContractNegotiationEventListener(eventRouter, clock));
+
         context.registerService(ContractNegotiationObservable.class, observable);
```
Comment (paullatzelsperger) [inline core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java:172]: not directly related, but maybe use `@Provider` methods?

**Review by paullatzelsperger — COMMENTED** [review]
I noticed that `ConfirmedContractNegotiationListener` and
`DeclinedContractNegotiationListener` are unused, so they can be removed. This then makes the `pre-`-methods of the `ContractNegotiationListener` obsolete as there no implementor left anymore.

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java (hunk 2 above)
Comment (ndr-brt) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java:172]: @paullatzelsperger as I specified in the last PR, the `@Provider` method does not work in certain conditions (at a first sight seems like you can't use injected services into a `@Provider` method, but I didn't dive into it):
https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/1587#discussion_r913881102

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/listener/ContractNegotiationEventListener.java (hunk 1 above)
Comment (ndr-brt) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/listener/ContractNegotiationEventListener.java]: you're right, I'm not fully aware of the difference between the two terms, anyway corrected :+1:

**Comment by ndr-brt** [comment]
@paullatzelsperger I think the `pre-` action methods are used by some implementors, as we discussed a lot about this especially with @algattik and @ouphi .
I would suggest to remove this `pre-` methods once the event system will get the persistence in a transactional context with the state change, doing that we're sure that every state change will have its own event (without the risk of losing them), this will enable audit on them. 
Just some ideas dropped here, we'd need to discuss if this is really needed and how it should be done.

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java (hunk 2 above)
Comment (paullatzelsperger) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java:172]: I guess the problem you're alluding to has to do with the sequence in which the phases of the DI process are executed. However, I managed to get this particular example to work (see [gist](https://gist.github.com/paullatzelsperger/5538411d2b269251c4acd8f0b7272c7d)).

**Comment by paullatzelsperger** [comment]
> ... the `pre-` action methods are used by some implementors, as we discussed a lot about this especially with @algattik and @ouphi .

There are none I could find in the EDC code base, but maybe there are some downstream. I'll have a look.

**Review by ronjaquensel — APPROVED** [review]

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java (hunk 2 above)
Comment (ndr-brt) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/ContractServiceExtension.java:172]: @paullatzelsperger ok, but the gist code moves to `@Provider` methods only stuff that's not related to the PR, I would address this change to another PR.

**Comment by paullatzelsperger** [comment]
> @paullatzelsperger I think the `pre-` action methods are used by some implementors, as we discussed a lot about this especially with @algattik and @ouphi . I would suggest to remove this `pre-` methods once the event system will get the persistence in a transactional context with the state change, doing that we're sure that every state change will have its own event (without the risk of losing them), this will enable audit on them. Just some ideas dropped here, we'd need to discuss if this is really needed and how it should be done.

@ndr-brt I looked at the code base of EDC, MVD and RegistrationService, found no usage the `pre-` methods of `ContractNegotiationListener`. I think we're good removing them. Is it possible you meant the `pre-` methods on the `TransferProcessListener`? There are indeed usages/implementations of those...

**Comment by ndr-brt** [comment]
> > @paullatzelsperger I think the `pre-` action methods are used by some implementors, as we discussed a lot about this especially with @algattik and @ouphi . I would suggest to remove this `pre-` methods once the event system will get the persistence in a transactional context with the state change, doing that we're sure that every state change will have its own event (without the risk of losing them), this will enable audit on them. Just some ideas dropped here, we'd need to discuss if this is really needed and how it should be done.
> 
> @ndr-brt I looked at the code base of EDC, MVD and RegistrationService, found no usage the `pre-` methods of `ContractNegotiationListener`. I think we're good removing them. Is it possible you meant the `pre-` methods on the `TransferProcessListener`? There are indeed usages/implementations of those...

Probably yes, I guess if there's the need for them in `TransferProcess` they could be needed on `ContractNegotiation` too.
Anyway will create an issue to remove them from the `ContractNegotiationListener`

**Review by paullatzelsperger — APPROVED** [review]

**Pull request merged**